File: application/core/src/main/java/com/aidriven/core/util/FileSummarizer.java
package com.aidriven.core.util;

import lombok.extern.slf4j.Slf4j;

import java.util.ArrayList;
import java.util.List;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

/**
 * Reduces large source files to structural summaries (signatures, declarations)
 * for inclusion in LLM context, targeting a 60–80% token reduction.
 *
 * <p>
 * Supports Java, TypeScript/JavaScript, and Python. All other file types
 * fall back to plain truncation with a notice. Files below the threshold are
 * returned unchanged.
 *
 * <p>
 * All methods are stateless and thread-safe.
 */
@Slf4j
public class FileSummarizer {

    private static final String TRUNCATION_NOTICE = "\n[... file truncated at summarization threshold ...]\n";

    private final int thresholdChars;

    public FileSummarizer(int thresholdChars) {
        this.thresholdChars = thresholdChars;
    }

    /**
     * Summarizes {@code content} if it exceeds the threshold.
     *
     * @param content   raw file content
     * @param extension file extension without the leading dot (e.g. "java", "ts")
     * @return summarized or original content, never null
     */
    public String summarize(String content, String extension) {
        if (content == null || content.isEmpty()) {
            return content == null ? "" : content;
        }
        if (content.length() <= thresholdChars) {
            return content;
        }

        String ext = extension != null ? extension.toLowerCase() : "";
        String summary = switch (ext) {
            case "java" -> summarizeJava(content);
            case "ts", "tsx" -> summarizeTypeScript(content);
            case "js", "jsx", "mjs" -> summarizeTypeScript(content); // same grammar
            case "py" -> summarizePython(content);
            default -> truncateWithNotice(content);
        };

        int originalLen = content.length();
        int summaryLen = summary.length();
        log.debug("FileSummarizer: .{} reduced {}->{} chars ({:.0f}% reduction)",
                ext, originalLen, summaryLen,
                summaryLen < originalLen ? 100.0 * (originalLen - summaryLen) / originalLen : 0.0);

        return summary;
    }

    // ─── Java ────────────────────────────────────────────────────────────────

    /**
     * Extracts:
     * - package declaration
     * - import statements (collapsed to first N)
     * - class/interface/enum/record declarations
     * - field declarations
     * - method/constructor signatures (first line only — no body)
     * - annotations on class/method level
     */
    private String summarizeJava(String content) {
        List<String> lines = List.of(content.split("\n", -1));
        StringBuilder sb = new StringBuilder();
        sb.append("[Java structural summary — method bodies omitted]\n\n");

        boolean inBlockComment = false;
        int braceDepth = 0;
        boolean inMethodBody = false;
        int methodBodyStartDepth = -1;

        // Patterns
        Pattern packagePat = Pattern.compile("^package\\s+[\\w.]+;");
        Pattern importPat = Pattern.compile("^import\\s+.*?;");
        Pattern classPat = Pattern.compile(
                "^(\\s*)(public|protected|private|abstract|final|sealed|static)?\\s*"
                        + "(class|interface|enum|record|@interface)\\s+\\w+");
        Pattern fieldPat = Pattern.compile(
                "^(\\s+)(public|protected|private|static|final|volatile|transient|"
                        + "readonly)?\\s*(static\\s+)?(final\\s+)?[\\w<>\\[\\].,?\\s]+ \\w+\\s*(=|;)");
        Pattern methodPat = Pattern.compile(
                "^(\\s+)(public|protected|private|static|default|abstract|final|"
                        + "synchronized|native|override|override)?\\s*(<[^>]+>\\s*)?"
                        + "[\\w<>\\[\\].,?\\s]+ \\w+\\s*\\([^)]*\\)");
        Pattern annotationPat = Pattern.compile("^\\s*@[A-Za-z]");

        int importCount = 0;
        String lastAnnotation = null;

        for (String line : lines) {
            String trimmed = line.trim();

            // Track block comments
            if (!inBlockComment && trimmed.startsWith("/*")) {
                inBlockComment = true;
            }
            if (inBlockComment) {
                if (trimmed.contains("*/"))
                    inBlockComment = false;
                continue; // skip comment lines
            }
            if (trimmed.startsWith("//"))
                continue; // skip line comments

            // Track brace depth
            long opens = trimmed.chars().filter(c -> c == '{').count();
            long closes = trimmed.chars().filter(c -> c == '}').count();

            // Package
            if (packagePat.matcher(trimmed).find()) {
                sb.append(line).append("\n");
                continue;
            }

            // Imports (keep first 20, then summarise remainder)
            if (importPat.matcher(trimmed).find()) {
                importCount++;
                if (importCount <= 20) {
                    sb.append(line).append("\n");
                } else if (importCount == 21) {
                    sb.append("// ... (additional imports omitted)\n");
                }
                continue;
            }

            // Annotations
            if (annotationPat.matcher(line).find()) {
                lastAnnotation = line;
                continue;
            }

            // Class/interface/enum declaration
            if (classPat.matcher(trimmed).find()) {
                if (lastAnnotation != null) {
                    sb.append(lastAnnotation).append("\n");
                }
                // Emit just the declaration line (stop at { or newline)
                String decl = line.contains("{") ? line.substring(0, line.lastIndexOf('{') + 1) : line;
                sb.append(decl).append("\n");
                lastAnnotation = null;
                braceDepth += (int) (opens - closes);
                continue;
            }

            // Update depth after processing struct declarations
            if (inMethodBody) {
                braceDepth += (int) (opens - closes);
                if (braceDepth <= methodBodyStartDepth) {
                    inMethodBody = false;
                    methodBodyStartDepth = -1;
                    sb.append("    }\n"); // close marker
                }
                continue; // skip body lines
            }

            // Field declarations (depth == 1 → class body)
            if (braceDepth == 1 && fieldPat.matcher(line).find()) {
                sb.append(line).append("\n");
                braceDepth += (int) (opens - closes);
                lastAnnotation = null;
                continue;
            }

            // Abstract method / interface method signature (ends in ';', no body)
            if (braceDepth >= 1 && trimmed.contains("(") && trimmed.endsWith(";")) {
                if (lastAnnotation != null) {
                    sb.append(lastAnnotation).append("\n");
                }
                sb.append(line).append("\n");
                lastAnnotation = null;
                braceDepth += (int) (opens - closes);
                continue;
            }

            // Method/constructor signature at class body depth (has body)
            if (braceDepth >= 1 && trimmed.contains("(") && methodPat.matcher(line).find() && trimmed.endsWith("{")) {
                if (lastAnnotation != null) {
                    sb.append(lastAnnotation).append("\n");
                }
                String sig = line.substring(0, line.lastIndexOf('{') + 1);
                sb.append(sig).append(" /* ... */ }\n");
                lastAnnotation = null;
                // The brace was opened → enter body to skip it
                methodBodyStartDepth = braceDepth;
                braceDepth += (int) (opens - closes);
                inMethodBody = braceDepth > methodBodyStartDepth;
                if (!inMethodBody)
                    methodBodyStartDepth = -1;
                continue;
            }

            // Closing braces
            if (trimmed.equals("}") || trimmed.equals("};")) {
                braceDepth = Math.max(0, braceDepth + (int) (opens - closes));
                sb.append(line).append("\n");
                continue;
            }

            // Everything else at depth 0 (annotations, etc.)
            braceDepth += (int) (opens - closes);
            lastAnnotation = null;
        }

        return sb.toString();
    }

    // ─── TypeScript / JavaScript ─────────────────────────────────────────────

    /**
     * Extracts:
     * - interface/type declarations
     * - export/class declarations
     * - function/method signatures (no body)
     * - import statements (first 15)
     */
    private String summarizeTypeScript(String content) {
        List<String> lines = List.of(content.split("\n", -1));
        StringBuilder sb = new StringBuilder();
        sb.append("[TypeScript/JS structural summary — function bodies omitted]\n\n");

        Pattern importPat = Pattern.compile("^(import|export\\s+\\{)");
        Pattern interfacePat = Pattern.compile("^(export\\s+)?(interface|type)\\s+\\w+");
        Pattern classPat = Pattern.compile("^(export\\s+)?(abstract\\s+)?class\\s+\\w+");
        Pattern funcPat = Pattern.compile(
                "^(export\\s+)?(async\\s+)?function\\s+\\w+\\s*\\(");
        Pattern arrowExportPat = Pattern.compile(
                "^export\\s+(const|let)\\s+\\w+\\s*=\\s*(async\\s+)?\\(");
        Pattern methodPat = Pattern.compile(
                "^\\s+(public|private|protected|static|async|get|set)?\\s*\\w+\\s*\\(");

        int importCount = 0;
        int braceDepth = 0;
        boolean inFuncBody = false;
        int funcBodyDepth = 0;

        for (String line : lines) {
            String trimmed = line.trim();
            if (trimmed.isEmpty())
                continue;
            if (trimmed.startsWith("//") || trimmed.startsWith("*"))
                continue;

            long opens = trimmed.chars().filter(c -> c == '{').count();
            long closes = trimmed.chars().filter(c -> c == '}').count();

            if (inFuncBody) {
                braceDepth += (int) (opens - closes);
                if (braceDepth <= funcBodyDepth) {
                    inFuncBody = false;
                    sb.append("}\n");
                }
                continue;
            }

            if (importPat.matcher(trimmed).find()) {
                importCount++;
                if (importCount <= 15)
                    sb.append(line).append("\n");
                else if (importCount == 16)
                    sb.append("// ... (imports omitted)\n");
                braceDepth += (int) (opens - closes);
                continue;
            }

            if (interfacePat.matcher(trimmed).find()) {
                // Emit entire interface block (interfaces are usually compact)
                sb.append(line).append("\n");
                braceDepth += (int) (opens - closes);
                continue;
            }

            if (classPat.matcher(trimmed).find() || funcPat.matcher(trimmed).find()
                    || arrowExportPat.matcher(trimmed).find()) {
                String sig = trimmed.contains("{")
                        ? line.substring(0, line.lastIndexOf('{') + 1)
                        : line;
                sb.append(sig);
                if (trimmed.contains("{")) {
                    sb.append(" /* ... */ }\n");
                    funcBodyDepth = braceDepth;
                    braceDepth += (int) (opens - closes);
                    inFuncBody = braceDepth > funcBodyDepth;
                } else {
                    sb.append("\n");
                    braceDepth += (int) (opens - closes);
                }
                continue;
            }

            if (methodPat.matcher(line).find() && braceDepth >= 1) {
                String sig = line.contains("{")
                        ? line.substring(0, line.lastIndexOf('{') + 1) + " /* ... */ }"
                        : line;
                sb.append(sig).append("\n");
                if (line.contains("{")) {
                    funcBodyDepth = braceDepth;
                    braceDepth += (int) (opens - closes);
                    inFuncBody = braceDepth > funcBodyDepth;
                }
                continue;
            }

            // Keep closing braces (class/type closings)
            if (trimmed.startsWith("}")) {
                sb.append(line).append("\n");
            }
            braceDepth += (int) (opens - closes);
        }

        return sb.toString();
    }

    // ─── Python ─────────────────────────────────────────────────────────────

    /**
     * Extracts:
     * - import statements (first 15)
     * - class declarations
     * - function/method signatures + one-line docstring
     */
    private String summarizePython(String content) {
        List<String> lines = List.of(content.split("\n", -1));
        StringBuilder sb = new StringBuilder();
        sb.append("[Python structural summary — function bodies omitted]\n\n");

        Pattern importPat = Pattern.compile("^(import|from)\\s+");
        Pattern classPat = Pattern.compile("^class\\s+\\w+");
        Pattern funcPat = Pattern.compile("^(    )?def\\s+\\w+\\s*\\(");
        Pattern docPat = Pattern.compile("\"\"\".*\"\"\"|'''.*'''|\"\"\".*|'''.*");

        int importCount = 0;
        boolean inFuncBody = false;
        int funcIndent = 0;
        boolean nextIsDocstring = false;

        for (String line : lines) {
            String trimmed = line.trim();

            if (importPat.matcher(trimmed).find()) {
                importCount++;
                if (importCount <= 15)
                    sb.append(line).append("\n");
                else if (importCount == 16)
                    sb.append("# ... (imports omitted)\n");
                continue;
            }

            if (classPat.matcher(trimmed).find()) {
                inFuncBody = false;
                sb.append(line).append("\n");
                continue;
            }

            if (funcPat.matcher(line).find()) {
                inFuncBody = true;
                funcIndent = leadingSpaces(line);
                nextIsDocstring = true;
                sb.append(line).append("\n");
                continue;
            }

            if (inFuncBody) {
                if (nextIsDocstring && (trimmed.startsWith("\"\"\"") || trimmed.startsWith("'''"))) {
                    sb.append(line).append("\n");
                    if (!trimmed.substring(3).contains("\"\"\"") && !trimmed.substring(3).contains("'''")) {
                        // multi-line docstring — find closing
                    }
                    nextIsDocstring = false;
                    continue;
                }
                nextIsDocstring = false;
                // Exit function body when indentation decreases
                if (!trimmed.isEmpty() && leadingSpaces(line) <= funcIndent) {
                    inFuncBody = false;
                    // Fall through to re-process this line
                } else {
                    continue; // skip body
                }
            }

            // De-indented line — check if it's a new function or class
            if (classPat.matcher(trimmed).find()) {
                sb.append(line).append("\n");
            } else if (funcPat.matcher(line).find()) {
                inFuncBody = true;
                funcIndent = leadingSpaces(line);
                nextIsDocstring = true;
                sb.append(line).append("\n");
            }
            // else: skip module-level code
        }

        return sb.toString();
    }

    private int leadingSpaces(String line) {
        int count = 0;
        for (char c : line.toCharArray()) {
            if (c == ' ')
                count++;
            else if (c == '\t')
                count += 4;
            else
                break;
        }
        return count;
    }

    // ─── Fallback ─────────────────────────────────────────────────────────────

    private String truncateWithNotice(String content) {
        return content.substring(0, thresholdChars) + TRUNCATION_NOTICE;
    }
}

File: go-app/internal/context/filesummarizer.go
package context

import (
	"fmt"
	"regexp"
	"strings"
)

// DefaultThreshold is the default character count below which files are returned unchanged.
const DefaultThreshold = 500

// FileSummarizer reduces large source files to structural summaries using regex-based extraction.
// This typically achieves 60-80% token reduction while preserving the structural overview.
type FileSummarizer struct {
	threshold int
}

// NewFileSummarizer creates a FileSummarizer with the given character threshold.
// Files with content length below the threshold are returned unchanged.
func NewFileSummarizer(threshold int) *FileSummarizer {
	if threshold <= 0 {
		threshold = DefaultThreshold
	}
	return &FileSummarizer{threshold: threshold}
}

// Summarize produces a structural summary of the given file content based on the file extension.
// Supported extensions: .java, .ts, .js, .tsx, .jsx, .py, .go.
// Unsupported extensions result in truncation with a notice.
func (f *FileSummarizer) Summarize(content, extension string) string {
	if len(content) <= f.threshold {
		return content
	}

	ext := strings.TrimPrefix(strings.ToLower(extension), ".")
	switch ext {
	case "java":
		return f.summarizeJava(content)
	case "ts", "tsx", "js", "jsx":
		return f.summarizeTypeScript(content)
	case "py":
		return f.summarizePython(content)
	case "go":
		return f.summarizeGo(content)
	default:
		return f.truncateUnsupported(content, ext)
	}
}

// Threshold returns the current character threshold.
func (f *FileSummarizer) Threshold() int {
	return f.threshold
}

// --- Java ---

var (
	javaPackageRe     = regexp.MustCompile(`(?m)^package\s+[^;]+;`)
	javaImportRe      = regexp.MustCompile(`(?m)^import\s+[^;]+;`)
	javaAnnotationRe  = regexp.MustCompile(`(?m)^@\w+(?:\([^)]*\))?`)
	javaClassDeclRe   = regexp.MustCompile(`(?m)^(?:public\s+|protected\s+|private\s+|abstract\s+|final\s+|static\s+)*(?:class|interface|enum|record)\s+\w+[^{]*`)
	javaFieldRe       = regexp.MustCompile(`(?m)^\s+(?:public|protected|private)\s+(?:static\s+)?(?:final\s+)?[\w<>\[\],\s]+\s+\w+\s*[;=]`)
	javaMethodSigRe   = regexp.MustCompile(`(?m)^\s+(?:@\w+(?:\([^)]*\))?\s+)*(?:public|protected|private)?\s*(?:static\s+)?(?:final\s+)?(?:synchronized\s+)?(?:abstract\s+)?[\w<>\[\],\s?]+\s+\w+\s*\([^)]*\)(?:\s*throws\s+[\w,\s]+)?`)
)

func (f *FileSummarizer) summarizeJava(content string) string {
	var sb strings.Builder
	sb.WriteString("// [SUMMARIZED - structural outline]\n\n")

	if m := javaPackageRe.FindString(content); m != "" {
		sb.WriteString(m)
		sb.WriteString("\n\n")
	}

	imports := javaImportRe.FindAllString(content, -1)
	if len(imports) > 0 {
		limit := min(len(imports), 20)
		for _, imp := range imports[:limit] {
			sb.WriteString(imp)
			sb.WriteString("\n")
		}
		if len(imports) > 20 {
			sb.WriteString(fmt.Sprintf("// ... and %d more imports\n", len(imports)-20))
		}
		sb.WriteString("\n")
	}

	annotations := javaAnnotationRe.FindAllString(content, -1)
	for _, ann := range annotations {
		sb.WriteString(ann)
		sb.WriteString("\n")
	}

	classDecls := javaClassDeclRe.FindAllString(content, -1)
	for _, decl := range classDecls {
		sb.WriteString(strings.TrimSpace(decl))
		sb.WriteString(" {\n")
	}

	fields := javaFieldRe.FindAllString(content, -1)
	if len(fields) > 0 {
		sb.WriteString("\n  // Fields\n")
		for _, field := range fields {
			sb.WriteString(strings.TrimRight(strings.TrimSpace(field), "="))
			sb.WriteString(";\n")
		}
	}

	methods := javaMethodSigRe.FindAllString(content, -1)
	if len(methods) > 0 {
		sb.WriteString("\n  // Methods\n")
		for _, method := range methods {
			sb.WriteString("  ")
			sb.WriteString(strings.TrimSpace(method))
			sb.WriteString(" { ... }\n")
		}
	}

	sb.WriteString("}\n")
	return sb.String()
}

// --- TypeScript / JavaScript ---

var (
	tsImportRe    = regexp.MustCompile(`(?m)^import\s+.+`)
	tsInterfaceRe = regexp.MustCompile(`(?m)^(?:export\s+)?(?:interface|type)\s+\w+[^{;]*`)
	tsClassDeclRe = regexp.MustCompile(`(?m)^(?:export\s+)?(?:abstract\s+)?class\s+\w+[^{]*`)
	tsFuncSigRe   = regexp.MustCompile(`(?m)^(?:export\s+)?(?:async\s+)?(?:function\s+\w+|(?:const|let|var)\s+\w+\s*=\s*(?:async\s+)?\([^)]*\)(?:\s*:\s*[\w<>\[\]|&\s]+)?\s*=>|(?:const|let|var)\s+\w+\s*=\s*(?:async\s+)?function)`)
)

func (f *FileSummarizer) summarizeTypeScript(content string) string {
	var sb strings.Builder
	sb.WriteString("// [SUMMARIZED - structural outline]\n\n")

	imports := tsImportRe.FindAllString(content, -1)
	if len(imports) > 0 {
		limit := min(len(imports), 15)
		for _, imp := range imports[:limit] {
			sb.WriteString(imp)
			sb.WriteString("\n")
		}
		if len(imports) > 15 {
			sb.WriteString(fmt.Sprintf("// ... and %d more imports\n", len(imports)-15))
		}
		sb.WriteString("\n")
	}

	interfaces := tsInterfaceRe.FindAllString(content, -1)
	for _, iface := range interfaces {
		sb.WriteString(strings.TrimSpace(iface))
		sb.WriteString(" { ... }\n")
	}

	classes := tsClassDeclRe.FindAllString(content, -1)
	for _, cls := range classes {
		sb.WriteString(strings.TrimSpace(cls))
		sb.WriteString(" { ... }\n")
	}

	funcs := tsFuncSigRe.FindAllString(content, -1)
	if len(funcs) > 0 {
		sb.WriteString("\n// Functions\n")
		for _, fn := range funcs {
			sb.WriteString(strings.TrimSpace(fn))
			sb.WriteString(" { ... }\n")
		}
	}

	return sb.String()
}

// --- Python ---

var (
	pyImportRe   = regexp.MustCompile(`(?m)^(?:import|from)\s+.+`)
	pyClassRe    = regexp.MustCompile(`(?m)^class\s+\w+[^:]*:`)
	pyFuncRe     = regexp.MustCompile(`(?m)^(?:    |\t)?(?:async\s+)?def\s+\w+\s*\([^)]*\)(?:\s*->\s*[\w\[\],\s|]+)?:`)
	pyDocstringRe = regexp.MustCompile(`(?m)^\s+"""([^"]*)"""`)
)

func (f *FileSummarizer) summarizePython(content string) string {
	var sb strings.Builder
	sb.WriteString("# [SUMMARIZED - structural outline]\n\n")

	imports := pyImportRe.FindAllString(content, -1)
	if len(imports) > 0 {
		limit := min(len(imports), 15)
		for _, imp := range imports[:limit] {
			sb.WriteString(imp)
			sb.WriteString("\n")
		}
		if len(imports) > 15 {
			sb.WriteString(fmt.Sprintf("# ... and %d more imports\n", len(imports)-15))
		}
		sb.WriteString("\n")
	}

	classes := pyClassRe.FindAllString(content, -1)
	for _, cls := range classes {
		sb.WriteString(strings.TrimSpace(cls))
		sb.WriteString("\n    ...\n")
	}

	lines := strings.Split(content, "\n")
	for i, line := range lines {
		if pyFuncRe.MatchString(line) {
			sb.WriteString(strings.TrimSpace(line))
			sb.WriteString("\n")
			// Look for docstring on the next few lines
			for j := i + 1; j < len(lines) && j <= i+3; j++ {
				trimmed := strings.TrimSpace(lines[j])
				if strings.HasPrefix(trimmed, `"""`) {
					if strings.Count(trimmed, `"""`) >= 2 {
						sb.WriteString("    ")
						sb.WriteString(trimmed)
						sb.WriteString("\n")
					} else {
						// Multi-line docstring: collect until closing """
						sb.WriteString("    ")
						sb.WriteString(trimmed)
						sb.WriteString("\n")
						for k := j + 1; k < len(lines) && k <= j+5; k++ {
							dtrimmed := strings.TrimSpace(lines[k])
							sb.WriteString("    ")
							sb.WriteString(dtrimmed)
							sb.WriteString("\n")
							if strings.Contains(dtrimmed, `"""`) {
								break
							}
						}
					}
					break
				}
				if trimmed != "" {
					break
				}
			}
			sb.WriteString("    ...\n\n")
		}
	}

	return sb.String()
}

// --- Go ---

var (
	goPackageRe  = regexp.MustCompile(`(?m)^package\s+\w+`)
	goImportRe   = regexp.MustCompile(`(?ms)^import\s*\(([^)]*)\)`)
	goImportOneRe = regexp.MustCompile(`(?m)^import\s+"[^"]*"`)
	goTypeDeclRe = regexp.MustCompile(`(?m)^type\s+\w+\s+(?:struct|interface|func)\b[^{]*`)
	goTypeAliasRe = regexp.MustCompile(`(?m)^type\s+\w+\s+\w+`)
	goFuncRe     = regexp.MustCompile(`(?m)^func\s+(?:\([^)]+\)\s+)?\w+\s*\([^)]*\)(?:\s*(?:\([^)]*\)|[\w*\[\]]+))?`)
)

func (f *FileSummarizer) summarizeGo(content string) string {
	var sb strings.Builder
	sb.WriteString("// [SUMMARIZED - structural outline]\n\n")

	if m := goPackageRe.FindString(content); m != "" {
		sb.WriteString(m)
		sb.WriteString("\n\n")
	}

	if m := goImportRe.FindStringSubmatch(content); len(m) > 1 {
		sb.WriteString("import (\n")
		sb.WriteString(m[1])
		sb.WriteString(")\n\n")
	} else {
		singles := goImportOneRe.FindAllString(content, -1)
		for _, s := range singles {
			sb.WriteString(s)
			sb.WriteString("\n")
		}
		if len(singles) > 0 {
			sb.WriteString("\n")
		}
	}

	types := goTypeDeclRe.FindAllString(content, -1)
	for _, t := range types {
		sb.WriteString(strings.TrimSpace(t))
		sb.WriteString(" { ... }\n")
	}

	aliases := goTypeAliasRe.FindAllString(content, -1)
	for _, a := range aliases {
		trimmed := strings.TrimSpace(a)
		// Skip those already captured as struct/interface/func
		if strings.Contains(trimmed, "struct") || strings.Contains(trimmed, "interface") || strings.Contains(trimmed, "func") {
			continue
		}
		sb.WriteString(trimmed)
		sb.WriteString("\n")
	}

	funcs := goFuncRe.FindAllString(content, -1)
	if len(funcs) > 0 {
		sb.WriteString("\n// Functions\n")
		for _, fn := range funcs {
			sb.WriteString(strings.TrimSpace(fn))
			sb.WriteString(" { ... }\n")
		}
	}

	return sb.String()
}

// --- Unsupported ---

const maxTruncateChars = 2000

func (f *FileSummarizer) truncateUnsupported(content, ext string) string {
	notice := fmt.Sprintf("// [TRUNCATED - .%s files are not supported for structural summarization]\n// Showing first %d characters.\n\n", ext, maxTruncateChars)
	if len(content) <= maxTruncateChars {
		return notice + content
	}
	return notice + content[:maxTruncateChars] + "\n// ... truncated"
}
